### ISSUE TITLE
MinItems in schema errors out for computed and optional fields

### ISSUE THREAD

**Issue by ravitandonrt** [hashicorp/terraform#19422]
### Terraform Version

```
2018/11/20 12:51:04 [INFO] Terraform version: 0.12.0 alpha2
2018/11/20 12:51:04 [INFO] Go runtime version: go1.11.1
```

### Terraform Configuration Files

```hcl
variable "autonomous_data_warehouse_cpu_core_count" {
  default = 1
}

variable "autonomous_data_warehouse_data_storage_size_in_tbs" {
  default = 1
}

variable "autonomous_data_warehouse_db_name" {
  default = "adwdb1"
}

variable "autonomous_data_warehouse_display_name" {
  default = "example_autonomous_data_warehouse"
}

variable "autonomous_data_warehouse_license_model" {
  default = "LICENSE_INCLUDED"
}

resource "random_string" "autonomous_data_warehouse_admin_password" {
  length  = 16
  special = true
}

resource "oci_database_autonomous_data_warehouse" "autonomous_data_warehouse" {
  #Required
  admin_password           = "${random_string.autonomous_data_warehouse_admin_password.result}"
  compartment_id           = "${var.compartment_ocid}"
  cpu_core_count           = "${var.autonomous_data_warehouse_cpu_core_count}"
  data_storage_size_in_tbs = "${var.autonomous_data_warehouse_data_storage_size_in_tbs}"
  db_name                  = "${var.autonomous_data_warehouse_db_name}"

  #Optional
  display_name  = "${var.autonomous_data_warehouse_display_name}"
  license_model = "${var.autonomous_data_warehouse_license_model}"
}
```

Resource schema link in [Additional Context](#Additional-Context)

### Debug Output

2018/11/20 12:51:06 [ERROR] <root>: eval: *terraform.EvalValidateResource, err: Insufficient connection_strings blocks: At least 1 "connection_strings" blocks are required.
2018/11/20 12:51:06 [ERROR] <root>: eval: *terraform.EvalSequence, err: Insufficient connection_strings blocks: At least 1 "connection_strings" blocks are required.
2018/11/20 12:51:06 [TRACE] [walkValidate] Exiting eval tree: oci_database_autonomous_data_warehouse.autonomous_data_warehouse
2018/11/20 12:51:06 [TRACE] vertex "oci_database_autonomous_data_warehouse.autonomous_data_warehouse": visit complete
2018/11/20 12:51:06 [TRACE] dag/walk: upstream of "output.parallel_connection_string" errored, so skipping

Error: Insufficient connection_strings blocks

  on /Users/ratandon/projects/go/src/github.com/terraform-providers/terraform-provider-oci/docs/examples/database/adw/autonomous_data_warehouse.tf line 17:
  17: }

At least 1 "connection_strings" blocks are required.

2018-11-20T12:51:06.781-0800 [WARN]  plugin: plugin failed to exit gracefully

### Expected Behavior

The previously compatible Terraform config should continue to function as is, without requiring to modify provider code to remove MinItems from optional parameters.

### Actual Behavior

Specifying the MinItems fails the validation step, requiring them to be specified even for fields that are marked computed or optional.

### Additional Context

Reference to connection_strings in resource schema: 
https://github.com/terraform-providers/terraform-provider-oci/blob/master/oci/database_autonomous_data_warehouse_resource.go#L80

Additionally, nested resources that involve TypeSets also seem to be impacted with this change where any items marked with MinItems require to be specified, for example: `icmp_options` in `ingress_security_rules` in `oci_security_list`

Reference: https://github.com/terraform-providers/terraform-provider-oci/blob/master/oci/core_security_list_resource.go#L64

**Comment by apparentlymart** [comment]
Hi @ravitandonrt!

Just to clarify: what was the intended meaning of setting this to `MinItems: 1, MaxItems: 1` if _not_ to consider it required?

**Comment by ravitandonrt** [comment]
Good point @apparentlymart I will find that information and update it here. It was likely done as part of swagger parsing to mark any arrays with Min and Max items based on what is specified. Looks like it has been wrongly evaluated in this case and defaulted to 1 if not present, but worked due to missing validations in previous terraform version.

**Comment by apparentlymart** [comment]
Hi again, @ravitandonrt! Thanks for sharing that additional information.

The new validation code considers the combination of "optional" and "minitems > 0" to be a nonsense case and so it isn't handling it properly. It sounds like what you'd like is for `Optional: true` to effectively cause `MinItems` to be ignored and assumed to be zero, which seems like something reasonable we can do to ensure backward-compatibility with how we formerly interpreted self-inconsistent schemas.

**Comment by apparentlymart** [comment]
This is fixed in #19478, by ignoring `MinItems` altogether when a nested block attribute (one whose `Elem` is a `schema.Resource`) is also marked as `Optional`.

For a nested block backed by a collection there is no situation where it can be both set and empty, since including one nested block (like `connection_strings`) both creates the collection _and_ populates it with its first element at the same time.

The other change in #19478 is to ignore `MinItems` and `MaxItems` altogether for an attribute that is `Computed: true, Optional: false`, since indeed these constraints have no meaning for a block that cannot be set in configuration.

Neither of these changes affect normal argument attributes (ones whose `Elem` is either unset or set to a `schema.Schema`), since for those `MinItems` and `MaxItems` are just additional validation rules checked by the provider code itself, rather than pre-validated by Terraform Core when decoding the configuration.
